### ISSUE TITLE
Not using a bounded generic type triggers a "cannot determine a type for this bounded type parameter" error.

### ISSUE THREAD

**Issue by miselin** [rust-lang/rust#15908]
This is a (very) contrived example that exhibits the error:
```
struct X {
    x: bool,
}

impl X {
    fn new() -> X {
        X{x: false}
    }
}

trait Thing {
    fn toggle(&mut self);
}

impl Thing for X {
    fn toggle(&mut self) {
        self.x = !self.x;
    }
}

fn bounded<T: Thing>(x: &mut X) {
    x.toggle()
}

fn main() {
    let mut x = X::new();
    bounded(&mut x);
}
```

Fires the error:
```
bounded.rs:28:5: 28:12 error: cannot determine a type for this bounded type parameter: unconstrained type
bounded.rs:28     bounded(&mut x);
                  ^~~~~~~
```

The error is fixed by redefining `bounded`:
```
fn bounded<T: Thing>(x: &mut T) { // Using 'T' instead of hardcoding the type.
    x.toggle()
}
```

Should this be caught by the linter?

**Comment by steveklabnik** [comment]
Triage: same error today.

**Comment by steveklabnik** [comment]
Triage: slightly different error now:

```
hello.rs:27:13: 27:20 error: unable to infer enough type information about `_`; type annotations or generic parameter binding required [E0282]
hello.rs:27             bounded(&mut x);
                        ^~~~~~~
hello.rs:27:13: 27:20 help: run `rustc --explain E0282` to see a detailed explanation
error: aborting due to previous error

```

**Comment by arielb1** [comment]
Does not look like a bug to me.
